lensless/utils/dataset.py

```diff
@@ -0,0 +1,443 @@
+        if self.transform_lensless:
+            lensless = self.transform_lensless(lensless)
+        if self.transform_lensed:
+            lensed = self.transform_lensed(lensed)
+
+        return lensless, lensed
+
+
+class SimulatedDataset(DualDataset):
+    """
+    Dataset of propagated images (through simulation) from a Torch Dataset.
+    The `waveprop <https://github.com/ebezzam/waveprop/blob/master/waveprop/simulation.py>`_ package is used for the simulation,
+    assuming a far-field propagation and a shift-invariant system with a single point spread function (PSF).
+
+    """
+
+    def __init__(
+        self,
+        psf,
+        dataset,
+        pre_transform=None,
+        dataset_is_CHW=False,
+        flip=False,
+        **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+
+        psf : torch.Tensor
+            PSF to use for simulate. Should be a 4D tensor with shape [1, H, W, C]. Simulation of multi-depth data is not supported yet.
+        dataset : torch.utils.data.Dataset
+            Dataset to propagate. Should output images with shape [H, W, C] unless ``dataset_is_CHW`` is ``True`` (and therefore images have the dimension ordering of [C, H, W]).
+        pre_transform : PyTorch Transform or None, optional
+            Transform to apply to the images before simulation, by default ``None``.
+        dataset_is_CHW : bool, optional
+            If True, the input dataset is expected to output images with shape [C, H, W], by default ``False``.
+        flip : bool, optional
+                If True,images are flipped beffore the simulation, by default ``False``..
+        """
+
+        # we do the flipping before the simualtion
+        super(SimulatedDataset, self).__init__(flip=False, **kwargs)
+
+        assert isinstance(dataset, Dataset)
+        self.dataset = dataset
+        self.n_files = len(dataset)
+
+        self.dataset_is_CHW = dataset_is_CHW
+        self._pre_transform = pre_transform
+        self.flip = flip
+
+        psf = psf.squeeze().movedim(-1, 0)
+
+        # initialize simulator
+        self.sim = FarFieldSimulator(psf=psf, is_torch=True, **kwargs)
+
+    def get_image(self, index):
+        return self.dataset[index]
+
+    def _get_images_pair(self, index):
+        # load image
+        img, _ = self.get_image(index)
+        if not self.dataset_is_CHW:
+            img = img.movedim(-1, 0)
+        if self.flip:
+            img = torch.rot90(img, dims=(-3, -2))
+        if self._pre_transform is not None:
+            img = self._pre_transform(img)
+
+        lensless, lensed = self.sim.propagate(img, return_object_plane=True)
+        lensless = lensless.moveaxis(-3, -1)
+        lensed = lensed.moveaxis(-3, -1)
+        return lensless, lensed
+
+    def __len__(self):
+        if self.indices is None:
+            return self.n_files
+        else:
+            return len([x for x in self.indices if x < self.n_files])
+
+
+class LenslessDataset(DualDataset):
```

Comment:
is this needed? seems identical to `ParallelDataset` and I don't see it used anywhere in the code

Reply:
This is for the case of the dataset you took in the lab. On of the image is a lensless measurement, and the other is the image that was projected on the screen. This is a bit different than `ParallelDataset` since we use the `farfieldsimulator` to project to object space. A similar result could probably be obtained using transform and `ParallelDataset` but this makes it a bit more straightforward.

Reply:
rename to ParallelDatasetSimulatedOriginal